### PR TITLE
CNV-7522 - nmstate more config example

### PR DESCRIPTION
**Pull request by aburdenthehand**
Adding some new nmstate configuration examples and snippets, which required a bit of rejigging of current content
And it looks like I name my branch the epic ID. Oh well. 

https://issues.redhat.com/browse/CNV-7522

Build link: https://cnv-7521-nmstate-examples--ocpdocs.netlify.app/openshift-enterprise/latest/virt/node_network/virt-updating-node-network-config.html

### CODE REVIEW

File: modules/virt-example-ethernet-nncp.adoc [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// Module included in the following assemblies:
+//
+// * virt/node_network/virt-updating-node-network-config.adoc
+
+[id="virt-example-ethernet-nncp_{context}"]
+= Example: Ethernet interface NodeNetworkConfigurationPolicy
+
+Create a ethernet interface on nodes in the cluster by applying a `NodeNetworkConfigurationPolicy` manifest
```
Comment (RamLavi) [inline modules/virt-example-ethernet-nncp.adoc:8]:
```suggestion
Create an ethernet interface on nodes in the cluster by applying a `NodeNetworkConfigurationPolicy` manifest
```

File: modules/virt-example-ethernet-nncp.adoc [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+to the cluster.
+
+The following YAML file is an example of a manifest for an ethernet interface.
+It includes samples values that you must replace with your own information.
+
+[source,yaml]
+----
+
+apiVersion: nmstate.io/v1beta1
```
Comment (RamLavi) [inline modules/virt-example-ethernet-nncp.adoc:17]: question: I noticed that we're using `nmstate.io/v1beta1` whereas in previous example `nmstate.io/v1alpha1` is used. should we align them (may be in a separate PR/commit of course)

File: modules/virt-example-ethernet-nncp.adoc [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+kind: NodeNetworkConfigurationPolicy
+metadata:
+  name: eth1-policy <1>
+spec:
+  nodeSelector: <2>
+    kubernetes.io/hostname: <node01> <3>
+  desiredState:
+    interfaces:
+    - name: eth1 <4>
+      description: Bond enslaving eth1 and eth2 <5>
```
Comment (RamLavi) [inline modules/virt-example-ethernet-nncp.adoc:27]: this description is not relevant to the type used

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+= Example snippet: Multiple interfaces in the same Policy
+
+You can create multiple interfaces in the same Policy. These interfaces can reference each other, allowing you to build and deploy a network configuration using a single Policy manifest.
+
+The following example snippet creates a bond (`bond10`) across two NICs and a Linux bridge (`br1`) that connects to the bond.
+
+[source,yaml]
+----
+...
+interfaces:
```
Comment (RamLavi) [inline modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:15]: I would keep the `description` field in all examples

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    dhcp: true
+    auto-gateway: false
+    auto-dns: false
+    enabled: true
+...
+----
+
+== DNS
+
+The following snippet creates an ethernet interface that
```
Comment (RamLavi) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:84]: that what? please complete the sentence

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    - example.com
+    - example.org
+    server:
+    - 8.8.8.8
+...
+----
+
+== Static routing
+
+The following snippet creates an ethernet interface that
```
Comment (RamLavi) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:103]: ditto

**Review by RamLavi — CHANGES_REQUESTED** [review]

File: modules/virt-example-snippets-nmstate-IP-management.adoc (hunk 6 above)
Comment (phoracek) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:103]: The following snipped configures a static route and a static IP on interfaces eth1. Note that the next-hop-address must be reachable through one of the interfaces.

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+[source,yaml]
+----
+...
+interfaces:
+  - name: eth1
+    type: ethernet
+    state: up
+    ipv4:
+      address:
+      - ip: 192.0.2.251
```
Comment (phoracek) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:114]: <1>

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        prefix-length: 24
+      enabled: true
+routes:
+  config:
+  - destination: 198.51.100.0/24
+    metric: 150
+    next-hop-address: 192.0.2.1
```
Comment (phoracek) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:121]: <2>

File: modules/virt-example-ethernet-nncp.adoc (hunk 2 above)
Comment (aburdenthehand) [reply modules/virt-example-ethernet-nncp.adoc:17]: Good catch. That update is scheduled for 2.6. Changing this to v1alpha1

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc (hunk 4 above)
Comment (aburdenthehand) [reply modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:15]: That's a good idea - done

**Review by yossisegev — APPROVED** [review]
Besides the single typo - LGTM

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc [hunk 9]
```diff
@@ -0,0 +1,32 @@
+// Module included in the following assemblies:
+//
+// * virt/node_network/virt-updating-node-network-config.adoc
+
+[id="virt-example-snippet-nmstate-multiple-interfaces_{context}"]
+= Example snippet: Multiple interfaces in the same Policy
```
Comment (sjhala-ccs) [inline modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:6]: Maybe remove the word "snippet" from the title, to be consistent with the other examples?

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+You can create multiple interfaces in the same Policy. These interfaces can reference each other, allowing you to build and deploy a network configuration using a single Policy manifest.
+
+The following example snippet creates a bond (`bond10`) across two NICs and a Linux bridge (`br1`) that connects to the bond.
```
Comment (sjhala-ccs) [inline modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:10]: I don't think that "snippet" is necessary when you're already saying "example".

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 11]
```diff
@@ -0,0 +1,132 @@
+// Module included in the following assemblies:
+//
+// * virt/node_network/virt-updating-node-network-config.adoc
+
+[id="virt-example-snippets-nmstate-IP-management_{context}"]
+= Example snippets: IP management
```
Comment (sjhala-ccs) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:6]: Remove "snippets" and change the title to "Examples: IP management"?

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+The following are example configuration snippets for different methods of IP management.
```
Comment (sjhala-ccs) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:8]: Maybe reword this as "The following are configuration examples for different methods of IP management"?

File: virt/node_network/virt-updating-node-network-config.adoc [hunk 13]
```diff
@@ -11,15 +11,31 @@ include::modules/virt-about-nmstate.adoc[leveloffset=+1]
 
 include::modules/virt-creating-interface-on-nodes.adoc[leveloffset=+1]
 
+[discrete]
```
Comment (sjhala-ccs) [inline virt/node_network/virt-updating-node-network-config.adoc:14]: Just curious, what is the purpose of the [discrete] label?

**Review by sjhala-ccs — COMMENTED** [review]
A few minor things, but otherwise LGTM!

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc (hunk 10 above)
Comment (aburdenthehand) [reply modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:10]: Despite what these changes may suggest, I actually don't much like the term 'snippet'. However I do think it's important to be very specific here between what is a complete example configuration (also featured in the assembly), and what is only a part of a configuration. nmstate is pretty powerful and these policies can alter the network of the whole cluster so I guess this is my prudence showing :)

File: virt/node_network/virt-updating-node-network-config.adoc (hunk 13 above)
Comment (aburdenthehand) [reply virt/node_network/virt-updating-node-network-config.adoc:14]: It prevents the title from showing in the TOC.
I used it here so that I could xref to the different policy examples after the early procedure where it's relevant, but having additional resources turn up as entry 3 in the TOC didn't look right (and those elements are already in the TOC, so it's only relevant to people who are looking at that particular procedure.

File: modules/virt-example-ethernet-nncp.adoc [hunk 14]
```diff
@@ -0,0 +1,41 @@
+// Module included in the following assemblies:
+//
+// * virt/node_network/virt-updating-node-network-config.adoc
+
+[id="virt-example-ethernet-nncp_{context}"]
+= Example: Ethernet interface NodeNetworkConfigurationPolicy
+
+Configure an ethernet interface on nodes in the cluster by applying a `NodeNetworkConfigurationPolicy` manifest to the cluster.
```
Comment (ousleyp) [inline modules/virt-example-ethernet-nncp.adoc:8]: s/ethernet/Ethernet for all instances of `ethernet` outside of YAMLs (ISG p 327)

File: modules/virt-example-ethernet-nncp.adoc [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+The following YAML file is an example of a manifest for an ethernet interface.
+It includes samples values that you must replace with your own information.
```
Comment (ousleyp) [inline modules/virt-example-ethernet-nncp.adoc:11]: s/samples/sample

File: modules/virt-example-ethernet-nncp.adoc [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+[source,yaml]
+----
+
```
Comment (ousleyp) [inline modules/virt-example-ethernet-nncp.adoc:15]: Extra newline?

File: modules/virt-example-ethernet-nncp.adoc [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    - name: eth1 <4>
+      description: Configuring eth1 on node01 <5>
+      type: ethernet <6>
+      state: up <7>
+      ipv4:
+        dhcp: true <8>
+        enabled: true <9>
+----
+<1> Name of the Policy.
+<2> Optional. If you do not include the `nodeSelector`, the Policy applies to all nodes in the cluster.
```
Comment (ousleyp) [inline modules/virt-example-ethernet-nncp.adoc:34]: I think the OCP docs moved from `Optional.` to `Optional:`

File: modules/virt-example-ethernet-nncp.adoc [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    interfaces:
+    - name: eth1 <4>
+      description: Configuring eth1 on node01 <5>
+      type: ethernet <6>
+      state: up <7>
+      ipv4:
+        dhcp: true <8>
+        enabled: true <9>
+----
+<1> Name of the Policy.
```
Comment (ousleyp) [inline modules/virt-example-ethernet-nncp.adoc:33]: Question that I have asked you before, but that might have a new answer now that the formatting guidelines have been updated: should Policy be capitalized throughout this document? If yes, should it be in backticks?

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc [hunk 19]
```diff
@@ -0,0 +1,32 @@
+// Module included in the following assemblies:
+//
+// * virt/node_network/virt-updating-node-network-config.adoc
+
+[id="virt-example-snippet-nmstate-multiple-interfaces_{context}"]
+= Example snippet: Multiple interfaces in the same Policy
+
+You can create multiple interfaces in the same Policy. These interfaces can reference each other, allowing you to build and deploy a network configuration using a single Policy manifest.
```
Comment (ousleyp) [inline modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:8]: s/using/by using

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc (hunk 10 above)
Comment (ousleyp) [inline modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:10]: I'd change these from parentheticals to something like:
```
...creates a bond that is named `bond10`...
```

File: modules/virt-example-snippets-nmstate-IP-management.adoc (hunk 11 above)
Comment (ousleyp) [reply modules/virt-example-snippets-nmstate-IP-management.adoc:6]: +1, I think that we can just say "Example" in the titles as long as we make it clear in the body of the module that this is a snippet.

File: modules/virt-example-snippets-nmstate-IP-management.adoc (hunk 12 above)
Comment (ousleyp) [reply modules/virt-example-snippets-nmstate-IP-management.adoc:8]: I agree this could be reworded, but I'd suggest something like "The following example configuration snippets demonstrate different methods of IP management."

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+// * virt/node_network/virt-updating-node-network-config.adoc
+
+[id="virt-example-snippets-nmstate-IP-management_{context}"]
+= Example snippets: IP management
+
+The following are example configuration snippets for different methods of IP management. 
+
+These snippets use the `ethernet` interface type to simplify the example while showing the related context in the Policy configuration. These IP management examples can be used with the other interface types. 
+
+== Static
```
Comment (ousleyp) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:12]: This needs an anchor.

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+      type: ethernet
+      state: up
+      ipv4:
+        address:
+        - ip: 192.168.122.250 <1>
+          prefix-length: 24
+        enabled: true
+...
+----
+<1> Replace with the static IP address for the interface.
```
Comment (ousleyp) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:31]: s/Replace with/Replace this value with

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+
+== No IP address
```
Comment (ousleyp) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:33]: ⚓  (and the same applies to other headings)

https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings

File: modules/virt-example-snippets-nmstate-IP-management.adoc [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+      description: DHCP on eth1
+      type: ethernet
+      state: up
+      ipv4:
+        dhcp: true
+        enabled: true
+...
+----
+
+The following snippet configures an ethernet interface that uses a dynamic IP address but without a dynamic gateway address or DNS:
```
Comment (ousleyp) [inline modules/virt-example-snippets-nmstate-IP-management.adoc:68]: s/but without/but does not use

**Review by ousleyp — COMMENTED** [review]
Some nits, but overall this is looking good!

File: modules/virt-example-ethernet-nncp.adoc (hunk 14 above)
Comment (aburdenthehand) [reply modules/virt-example-ethernet-nncp.adoc:8]: Great catch, that passed me by completely

File: modules/virt-example-ethernet-nncp.adoc (hunk 17 above)
Comment (aburdenthehand) [reply modules/virt-example-ethernet-nncp.adoc:34]: In that case I've updated all the instances in this assembly

File: modules/virt-example-ethernet-nncp.adoc (hunk 18 above)
Comment (aburdenthehand) [reply modules/virt-example-ethernet-nncp.adoc:33]: Good question. Not in backticks, as 'Policy' is only part of the full object name. I'm reluctant to drop it down to 'policy' though because we also have other policies that impact VMs. Maybe using the acronym (NNCP) makes the most sense but runs into the danger of blurring with NNCE, and that distinction should remain clear. 
I'd like to keep capitalised for now as I think that makes the most sense for now, but I've raised an item for our next meeting for addressing the decapitalisation of objects in our docs en masse.

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc (hunk 9 above)
Comment (aburdenthehand) [reply modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:6]: Updated titles, ids, and filenames to drop snippet for both offending modules.

File: modules/virt-example-snippets-nmstate-IP-management.adoc (hunk 11 above)
Comment (aburdenthehand) [reply modules/virt-example-snippets-nmstate-IP-management.adoc:6]: Dropped from both files

File: modules/virt-example-snippets-nmstate-IP-management.adoc (hunk 22 above)
Comment (aburdenthehand) [reply modules/virt-example-snippets-nmstate-IP-management.adoc:33]: Aye aye!

File: modules/virt-example-snippets-nmstate-IP-management.adoc (hunk 12 above)
Comment (aburdenthehand) [reply modules/virt-example-snippets-nmstate-IP-management.adoc:8]: :alf:

File: modules/virt-example-snippet-nmstate-multiple-interfaces.adoc (hunk 10 above)
Comment (aburdenthehand) [reply modules/virt-example-snippet-nmstate-multiple-interfaces.adoc:10]: Cleaned up some of the flagrant usage. Now only used in leading para specific to the config
